File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  defp is_yelling(input) do
    String.upcase(input) == input && String.downcase(input) !== input
  end

  defp is_question(input) do
    input
    |> String.trim()
    |> String.ends_with?("?")
  end

  defp is_silence(input) do
    String.replace(input, ~r/[\r\t\n]/, "") == ""
  end

  def hey(input) do
    input = String.trim(input)

    cond do
      is_silence(input) -> "Fine. Be that way!"
      is_yelling(input) && is_question(input) -> "Calm down, I know what I'm doing!"
      is_yelling(input) -> "Whoa, chill out!"
      is_question(input) -> "Sure."
      true -> "Whatever."
    end
  end
end

File: solutions/fsharp/bob/4/Bob.fs
module Bob

let (|Question|_|) (input: string) =
    if input.EndsWith('?') then Some() else None

let (|Yelling|_|) (input: string) =
    if input <> input.ToLower() && input = input.ToUpper() then
        Some()
    else
        None

let response (input: string) =
    match input.Trim() with
    | Question & Yelling -> "Calm down, I know what I'm doing!"
    | Question -> "Sure."
    | Yelling -> "Whoa, chill out!"
    | "" -> "Fine. Be that way!"
    | _ -> "Whatever."
